File: android-im-ui/src/main/kotlin/com/flare/im/ui/IconLibrary.kt
package com.flare.im.ui

import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material.icons.automirrored.outlined.HelpOutline
import androidx.compose.material.icons.automirrored.outlined.Forward
import androidx.compose.material.icons.automirrored.outlined.Logout
import androidx.compose.material.icons.automirrored.outlined.Reply
import androidx.compose.material.icons.automirrored.outlined.Send
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material.icons.outlined.ArrowDownward
import androidx.compose.material.icons.outlined.BookmarkBorder
import androidx.compose.material.icons.outlined.CalendarToday
import androidx.compose.material.icons.outlined.Call
import androidx.compose.material.icons.outlined.Cancel
import androidx.compose.material.icons.outlined.ChatBubbleOutline
import androidx.compose.material.icons.outlined.Check
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.ChevronRight
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.ContentCopy
import androidx.compose.material.icons.outlined.Delete
import androidx.compose.material.icons.outlined.Description
import androidx.compose.material.icons.outlined.Download
import androidx.compose.material.icons.outlined.Edit
import androidx.compose.material.icons.outlined.Block
import androidx.compose.material.icons.outlined.Campaign
import androidx.compose.material.icons.outlined.DarkMode
import androidx.compose.material.icons.outlined.Devices
import androidx.compose.material.icons.outlined.EmojiEmotions
import androidx.compose.material.icons.outlined.Explore
import androidx.compose.material.icons.outlined.Label
import androidx.compose.material.icons.outlined.Language
import androidx.compose.material.icons.outlined.ExpandMore
import androidx.compose.material.icons.outlined.FavoriteBorder
import androidx.compose.material.icons.outlined.Folder
import androidx.compose.material.icons.outlined.Forum
import androidx.compose.material.icons.outlined.Image
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material.icons.outlined.Link
import androidx.compose.material.icons.outlined.LocationOn
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material.icons.outlined.Mic
import androidx.compose.material.icons.outlined.MoreHoriz
import androidx.compose.material.icons.outlined.Notifications
import androidx.compose.material.icons.outlined.NotificationsOff
import androidx.compose.material.icons.outlined.People
import androidx.compose.material.icons.outlined.Person
import androidx.compose.material.icons.outlined.PersonAddAlt
import androidx.compose.material.icons.outlined.PhotoCamera
import androidx.compose.material.icons.outlined.QrCode
import androidx.compose.material.icons.outlined.PushPin
import androidx.compose.material.icons.outlined.Refresh
import androidx.compose.material.icons.outlined.Remove
import androidx.compose.material.icons.outlined.Schedule
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material.icons.outlined.Settings
import androidx.compose.material.icons.outlined.Share
import androidx.compose.material.icons.outlined.StarBorder
import androidx.compose.material.icons.outlined.Videocam
import androidx.compose.material.icons.outlined.Visibility
import androidx.compose.material.icons.outlined.VisibilityOff
import androidx.compose.material.icons.outlined.WarningAmber
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * Cross-platform icon-library composable. Maps a fixed set of 60 semantic
 * names to the closest Material [ImageVector]s so every Flare platform can
 * share one naming contract.
 */
private val flareIconMap: Map<String, ImageVector> = mapOf(
    "search" to Icons.Outlined.Search,
    "send" to Icons.AutoMirrored.Outlined.Send,
    "more" to Icons.Outlined.MoreHoriz,
    "back" to Icons.AutoMirrored.Outlined.ArrowBack,
    "close" to Icons.Outlined.Close,
    "check" to Icons.Outlined.Check,
    "add" to Icons.Outlined.Add,
    "remove" to Icons.Outlined.Remove,
    "edit" to Icons.Outlined.Edit,
    "delete" to Icons.Outlined.Delete,
    "heart" to Icons.Outlined.FavoriteBorder,
    "heart-filled" to Icons.Filled.Favorite,
    "comment" to Icons.Outlined.ChatBubbleOutline,
    "share" to Icons.Outlined.Share,
    "camera" to Icons.Outlined.PhotoCamera,
    "image" to Icons.Outlined.Image,
    "location" to Icons.Outlined.LocationOn,
    "mic" to Icons.Outlined.Mic,
    "phone" to Icons.Outlined.Call,
    "video" to Icons.Outlined.Videocam,
    "settings" to Icons.Outlined.Settings,
    "person" to Icons.Outlined.Person,
    "people" to Icons.Outlined.People,
    "person-add" to Icons.Outlined.PersonAddAlt,
    "star" to Icons.Outlined.StarBorder,
    "bookmark" to Icons.Outlined.BookmarkBorder,
    "download" to Icons.Outlined.Download,
    "link" to Icons.Outlined.Link,
    "emoji" to Icons.Outlined.EmojiEmotions,
    "file" to Icons.Outlined.Description,
    "folder" to Icons.Outlined.Folder,
    "notification" to Icons.Outlined.Notifications,
    "mute" to Icons.Outlined.NotificationsOff,
    "copy" to Icons.Outlined.ContentCopy,
    "forward" to Icons.AutoMirrored.Outlined.Forward,
    "reply" to Icons.AutoMirrored.Outlined.Reply,
    "refresh" to Icons.Outlined.Refresh,
    "chevron-down" to Icons.Outlined.ExpandMore,
    "chevron-right" to Icons.Outlined.ChevronRight,
    "arrow-down" to Icons.Outlined.ArrowDownward,
    "warning" to Icons.Outlined.WarningAmber,
    "info" to Icons.Outlined.Info,
    "success" to Icons.Outlined.CheckCircle,
    "error" to Icons.Outlined.Cancel,
    "calendar" to Icons.Outlined.CalendarToday,
    "clock" to Icons.Outlined.Schedule,
    "eye" to Icons.Outlined.Visibility,
    "eye-off" to Icons.Outlined.VisibilityOff,
    "lock" to Icons.Outlined.Lock,
    "qr" to Icons.Outlined.QrCode,
    "chats" to Icons.Outlined.Forum,
    "moments" to Icons.Outlined.Explore,
    "block" to Icons.Outlined.Block,
    "tag" to Icons.Outlined.Label,
    "announcement" to Icons.Outlined.Campaign,
    "theme" to Icons.Outlined.DarkMode,
    "language" to Icons.Outlined.Language,
    "devices" to Icons.Outlined.Devices,
    "logout" to Icons.AutoMirrored.Outlined.Logout,
    "pin" to Icons.Outlined.PushPin,
)

/** The 60 semantic icon names, in canonical order. */
val flareIconNames: List<String> = listOf(
    "search", "send", "more", "back", "close", "check", "add", "remove",
    "edit", "delete", "heart", "heart-filled", "comment", "share", "camera",
    "image", "location", "mic", "phone", "video", "settings", "person",
    "people", "person-add", "star", "bookmark", "download", "link", "emoji",
    "file", "folder", "notification", "mute", "copy", "forward", "reply",
    "refresh", "chevron-down", "chevron-right", "arrow-down", "warning",
    "info", "success", "error", "calendar", "clock", "eye", "eye-off",
    "lock", "qr", "chats", "moments", "block", "tag", "announcement",
    "theme", "language", "devices", "logout",
    "pin",
)

/** Resolve a semantic icon [name] to a Material [ImageVector]; unknown → HelpOutline. */
fun flareIconVector(name: String): ImageVector =
    flareIconMap[name] ?: Icons.AutoMirrored.Outlined.HelpOutline

/** Render a Flare semantic icon by [name]. */
@Composable
fun FlareIcon(name: String, size: Dp = 20.dp, tint: Color? = null) {
    Icon(
        imageVector = flareIconVector(name),
        contentDescription = name,
        tint = tint ?: flareColors().textSecondary,
        modifier = Modifier.size(size),
    )
}

File: packages/ios-im-ui/Sources/FlareIMUI/Components/IconLibrary.swift
import SwiftUI

/// Ordered list of the 105 semantic icon names supported by `IconView`.
public let flareIconNames: [String] = [
    "search", "send", "more", "back", "close",
    "check", "add", "remove", "edit", "delete",
    "heart", "comment", "chats", "moments", "share", "camera",
    "image", "location", "mic", "phone", "video",
    "settings", "person", "people", "person-add", "star",
    "download", "link", "emoji", "file",
    "folder", "notification", "mute", "copy", "forward",
    "reply", "refresh", "chevron-down", "chevron-right", "arrow-down",
    "warning", "info", "success", "error", "calendar",
    "clock", "eye", "eye-off", "lock", "qr",
    "block", "tag", "announcement", "theme", "language", "devices", "logout",
    "pin", "poll",
    // Message actions
    "recall", "unpin", "merge-forward", "multi-select", "quote",
    "reaction", "translate", "mention", "read", "mark",
    // Composer
    "rich-text", "attachment",
    // Conversation
    "mark-unread", "archive", "unarchive", "clear-history",
    // Calls
    "mic-off", "camera-off", "speaker", "speaker-off", "end-call",
    "switch-camera", "screen-share",
    // Media
    "play", "pause", "expand", "collapse", "zoom-in", "zoom-out", "rotate",
    // Members
    "group", "admin", "remove-member", "transfer-owner", "silence", "report",
    // General
    "chevron-up", "chevron-left", "keyboard", "mini-app",
    "pin-self", "diagnostics", "card", "id", "join-request", "storage",
]

/// Maps each semantic icon name to the closest SF Symbol name. Every symbol is in the SF Symbols
/// catalogue at the package's iOS 16 / macOS 13 floor; where the natural symbol is newer or does not
/// exist, the entry names the closest symbol with the same concept.
public let flareIconMap: [String: String] = [
    "search": "magnifyingglass",
    "send": "paperplane",
    "more": "ellipsis",
    "back": "chevron.left",
    "close": "xmark",
    "check": "checkmark",
    "add": "plus",
    "remove": "minus",
    "edit": "pencil",
    "delete": "trash",
    "heart": "heart",
    "comment": "bubble.left",
    "share": "square.and.arrow.up",
    "camera": "camera",
    "image": "photo",
    // A map pin, as the location message and attach tile draw it (`location` is the current-position arrow).
    "location": "mappin.and.ellipse",
    "mic": "mic",
    "phone": "phone",
    "video": "video",
    "settings": "gearshape",
    "person": "person",
    "people": "person.2",
    "person-add": "person.badge.plus",
    "star": "star",
    "download": "arrow.down.circle",
    "link": "link",
    "emoji": "face.smiling",
    "file": "doc",
    "folder": "folder",
    "notification": "bell",
    "mute": "bell.slash",
    "copy": "doc.on.doc",
    "forward": "arrowshape.turn.up.right",
    "reply": "arrowshape.turn.up.left",
    "refresh": "arrow.clockwise",
    "chevron-down": "chevron.down",
    "chevron-right": "chevron.right",
    "arrow-down": "arrow.down",
    "warning": "exclamationmark.triangle",
    "info": "info.circle",
    "success": "checkmark.circle",
    // An exclamation circle: an x-circle reads as cancel.
    "error": "exclamationmark.circle",
    "calendar": "calendar",
    "clock": "clock",
    "eye": "eye",
    "eye-off": "eye.slash",
    "lock": "lock",
    "qr": "qrcode",
    "chats": "bubble.left.and.bubble.right",
    "moments": "safari",
    "block": "nosign",
    "tag": "tag",
    "announcement": "megaphone",
    "theme": "moon",
    "language": "globe",
    // More than one device, not a single laptop.
    "devices": "laptopcomputer.and.iphone",
    "logout": "rectangle.portrait.and.arrow.right",
    "pin": "pin",
    "poll": "chart.bar",
    "recall": "arrow.uturn.backward",
    "unpin": "pin.slash",
    "merge-forward": "arrow.triangle.merge",
    "multi-select": "checklist",
    "quote": "quote.opening",
    // SF Symbols has no face-with-plus; the add-reaction key shares the emoji face.
    "reaction": "face.smiling",
    // `translate` needs iOS 17.4; `character.bubble` is the translation glyph available at iOS 16.
    "translate": "character.bubble",
    "mention": "at",
    // SF Symbols has no double check; an opened envelope reads as "read".
    "read": "envelope.open",
    "mark": "flag",
    "rich-text": "textformat",
    "attachment": "paperclip",
    "mark-unread": "message.badge",
    "archive": "archivebox",
    "unarchive": "tray.and.arrow.up",
    "clear-history": "eraser",
    "mic-off": "mic.slash",
    "camera-off": "video.slash",
    "speaker": "speaker.wave.2",
    "speaker-off": "speaker.slash",
    "end-call": "phone.down",
    "switch-camera": "arrow.triangle.2.circlepath.camera",
    "screen-share": "rectangle.on.rectangle",
    "play": "play",
    "pause": "pause",
    "expand": "arrow.up.left.and.arrow.down.right",
    "collapse": "arrow.down.right.and.arrow.up.left",
    "zoom-in": "plus.magnifyingglass",
    "zoom-out": "minus.magnifyingglass",
    "rotate": "rotate.right",
    "group": "person.3",
    "admin": "checkmark.shield",
    "remove-member": "person.badge.minus",
    "transfer-owner": "key",
    // A member who cannot speak. SF Symbols has no slashed speech bubble; a silenced voice waveform
    // is the same concept and stays apart from `speaker-off` (audio output) and `mic-off`.
    "silence": "waveform.slash",
    "report": "exclamationmark.octagon",
    "chevron-up": "chevron.up",
    "chevron-left": "chevron.left",
    "keyboard": "keyboard",
    "mini-app": "square.grid.2x2",
    // Round 9: concepts every kit drew under a name that meant something else.
    "pin-self": "bookmark",
    "diagnostics": "ladybug",
    "card": "person.crop.rectangle",
    "id": "number",
    "join-request": "tray.and.arrow.down",
    "storage": "internaldrive",
]

/// The SF Symbol drawn for a semantic icon name; an unknown name draws `questionmark`. The one
/// resolution every kit icon name goes through (``IconView``, header actions, ``ActionMenuView`` items,
/// and every `icon:` the kit's models take).
///
/// A name that is not in the registry draws the fallback and says so once in a debug build — an SF
/// Symbol name passed where a kit name belongs (`"person.2"` for `people`) is not a compile error and
/// would otherwise be a silent question mark in the product.
func flareIconSymbol(_ name: String) -> String {
    if let symbol = flareIconMap[name] { return symbol }
    #if DEBUG
    flareWarnUnknownIcon(name)
    #endif
    return "questionmark"
}

#if DEBUG
private let flareUnknownIconLock = NSLock()
private nonisolated(unsafe) var flareWarnedIcons: Set<String> = []

/// Says an icon name is unknown once per name, so a list of 200 rows does not print 200 lines.
private func flareWarnUnknownIcon(_ name: String) {
    flareUnknownIconLock.lock()
    let unseen = flareWarnedIcons.insert(name).inserted
    flareUnknownIconLock.unlock()
    guard unseen else { return }
    print("[FlareIMUI] unknown icon name \"\(name)\" — pass one of flareIconNames; drawing the fallback glyph")
}
#endif

/// Action ids that stand for a kit icon name, so a header action needs no `icon` of its own to draw
/// the right glyph. The same four on every platform.
let flareActionIconAliases: [String: String] = [
    "audioCall": "phone",
    "videoCall": "video",
    "addMember": "person-add",
    "details": "info",
]

/// The SF Symbol for an action's icon name: an alias in ``flareActionIconAliases`` first, then the
/// kit vocabulary; an unknown name draws `questionmark`, never `more`.
func flareActionSymbol(_ name: String) -> String {
    flareIconSymbol(flareActionIconAliases[name] ?? name)
}

/// Whether an action id names an icon by itself: an alias or a kit icon name.
func flareActionIdNamesIcon(_ id: String) -> Bool {
    flareActionIconAliases[id] != nil || flareIconMap[id] != nil
}

/// Renders a Flare semantic icon as an SF Symbol.
///
/// Unknown names fall back to `questionmark`. The default colour is the
/// theme-aware `textSecondary` token.
public struct IconView: View {
    private let name: String
    private let size: CGFloat
    private let color: Color?
    private let accessibilityLabel: String?

    @Environment(\.colorScheme) private var scheme
    @Environment(\.flareBrandTheme) private var flareBrandTheme

    /// `accessibilityLabel` is the icon's own name. Leave it nil for a decorative
    /// icon inside an already-labelled control: the symbol is hidden from
    /// VoiceOver instead of announcing SF Symbols' own name over that label.
    public init(_ name: String, size: CGFloat = 20, color: Color? = nil, accessibilityLabel: String? = nil) {
        self.name = name
        self.size = size
        self.color = color
        self.accessibilityLabel = accessibilityLabel
    }

    public var body: some View {
        Image(systemName: flareIconSymbol(name))
            .font(.system(size: size))
            .foregroundColor(color ?? FlareColors.of(scheme, brand: flareBrandTheme).textSecondary)
            .accessibilityLabel(accessibilityLabel ?? "")
            .accessibilityHidden(accessibilityLabel == nil)
    }
}
